File: scripts/backup-db.ts
import 'dotenv/config';
import { db } from '../src/db/index.ts';
import fs from 'fs';
import path from 'path';

import {
  organizations,
  roles,
  permissions,
  users,
  donors,
  projects,
  projectMembers,
  agreements,
  disbursements,
  clauses,
  budgetVersions,
  budgetLines,
  expenses,
  receiptsVouchers,
  documents,
  reports,
  indicators,
  auditLogs,
  notifications,
  tasks,
  taskDependencies,
  taskComments,
  projectLogs,
  events,
  eventAttendees,
  documentAnalysis
} from '../src/db/schema.ts';

async function backup() {
  console.log('Iniciando respaldo de base de datos local...');
  const backupData: any = {};

  try {
    backupData.organizations = await db.select().from(organizations);
    backupData.roles = await db.select().from(roles);
    backupData.permissions = await db.select().from(permissions);
    backupData.users = await db.select().from(users);
    backupData.donors = await db.select().from(donors);
    backupData.projects = await db.select().from(projects);
    backupData.projectMembers = await db.select().from(projectMembers);
    backupData.agreements = await db.select().from(agreements);
    backupData.disbursements = await db.select().from(disbursements);
    backupData.clauses = await db.select().from(clauses);
    backupData.budgetVersions = await db.select().from(budgetVersions);
    backupData.budgetLines = await db.select().from(budgetLines);
    backupData.expenses = await db.select().from(expenses);
    backupData.receiptsVouchers = await db.select().from(receiptsVouchers);
    backupData.documents = await db.select().from(documents);
    backupData.reports = await db.select().from(reports);
    backupData.indicators = await db.select().from(indicators);
    backupData.auditLogs = await db.select().from(auditLogs);
    backupData.notifications = await db.select().from(notifications);
    backupData.tasks = await db.select().from(tasks);
    backupData.taskDependencies = await db.select().from(taskDependencies);
    backupData.taskComments = await db.select().from(taskComments);
    backupData.projectLogs = await db.select().from(projectLogs);
    backupData.events = await db.select().from(events);
    backupData.eventAttendees = await db.select().from(eventAttendees);
    backupData.documentAnalysis = await db.select().from(documentAnalysis);

    const dateStr = new Date().toISOString().replace(/[:.]/g, '-');
    const filename = `respaldo_proyecty_${dateStr}.json`;
    const filepath = path.join(process.cwd(), filename);
    
    fs.writeFileSync(filepath, JSON.stringify(backupData, null, 2), 'utf-8');
    
    console.log(`\n¡Respaldo completado con éxito!`);
    console.log(`Archivo guardado en: ${filepath}`);
    console.log(`Tablas respaldadas: ${Object.keys(backupData).length}`);
    
  } catch (error) {
    console.error('Error durante el respaldo:', error);
  } finally {
    process.exit(0);
  }
}

backup();

File: scripts/backup-db.py
import urllib.request
import json
import datetime
import os

SUPABASE_URL = "https://kwmvuuwinufksjjfsuls.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6Imt3bXZ1dXdpbnVma3NqamZzdWxzIiwicm9sZSI6InNlcnZpY2Vfcm9sZSIsImlhdCI6MTc4MjMyODM2OSwiZXhwIjoyMDk3OTA0MzY5fQ.0Q_wVVgFAF4rXIZxG2BTbY-dkjwYuY-WEGwxP7r7ygE"

TABLES = [
    "organizations", "roles", "permissions", "users", "donors", "projects", 
    "project_members", "agreements", "disbursements", "clauses", "budget_versions", 
    "budget_lines", "expenses", "receipts_vouchers", "documents", "reports", 
    "indicators", "audit_logs", "notifications", "tasks", "task_dependencies", 
    "task_comments", "project_logs", "events", "event_attendees", "document_analysis"
]

backup_data = {}

print("Iniciando respaldo de base de datos desde Supabase...")

for table in TABLES:
    req = urllib.request.Request(
        f"{SUPABASE_URL}/rest/v1/{table}?select=*",
        headers={
            "apikey": SUPABASE_KEY,
            "Authorization": f"Bearer {SUPABASE_KEY}"
        }
    )
    try:
        with urllib.request.urlopen(req) as response:
            data = json.loads(response.read().decode())
            backup_data[table] = data
            print(f"Respaldada tabla: {table} ({len(data)} registros)")
    except Exception as e:
        print(f"Error al respaldar tabla {table}: {e}")

date_str = datetime.datetime.now().strftime("%Y-%m-%dT%H-%M-%S")
filename = f"respaldo_proyecty_{date_str}.json"
filepath = os.path.join(os.getcwd(), filename)

with open(filepath, "w", encoding="utf-8") as f:
    json.dump(backup_data, f, indent=2, ensure_ascii=False)

print(f"\n¡Respaldo completado con éxito!")
print(f"Archivo guardado en: {filepath}")
